crates/table/src/bflatn_to_bsatn_fast_path.rs

```diff
@@ -0,0 +1,424 @@
+        builder.visit_product(row_type.product())?;
+        Some(builder.build())
+    }
+}
+
+/// An identifier for a series of bytes within a BFLATN row
+/// which can be directly copied into an output BSATN buffer
+/// with a known length and offset.
+///
+/// Within the row type's BFLATN layout, `row[bflatn_offset .. (bflatn_offset + length)]`
+/// must not contain any padding bytes,
+/// i.e. all of those bytes must be fully initialized if the row is initialized.
+#[derive(PartialEq, Eq, Debug)]
+struct MemcpyField {
+    /// Offset in the BFLATN row from which to begin `memcpy`ing, in bytes.
+    bflatn_offset: u16,
+
+    /// Offset in the BSATN buffer to which to begin `memcpy`ing, in bytes.
+    // TODO(comment): Could be a running counter, but this way we just have all the `memcpy` args in one place.
+    // Should bench; I suspect this allows more insn parallelism and is therefore better.
+    bsatn_offset: u16,
+
+    /// Length to `memcpy`, in bytes.
+    length: u16,
+}
+
+impl MemcpyField {
+    /// # Safety
+    ///
+    /// - `buf` must be at least `self.bsatn_offset + self.length` long.
+    /// - `row` must be at least `self.bflatn_offset + self.length` long.
+    /// - `row[self.bflatn_offset .. self.bflatn_offset + length]` must all be initialized.
+    unsafe fn copy(&self, buf: &mut [u8], row: &Bytes) {
+        // Safety: forward caller requirement #1.
+        let to = unsafe { buf.get_unchecked_mut(range_move(0..self.length as usize, self.bsatn_offset as usize)) };
+        // Safety: forward caller requirement #2.
+        let from = unsafe { row.get_unchecked(range_move(0..self.length as usize, self.bflatn_offset as usize)) };
+        // Safety: forward caller requirement #3.
+        let from = unsafe { slice_assume_init_ref(from) };
+        to.copy_from_slice(from);
+    }
+
+    fn is_empty(&self) -> bool {
+        self.length == 0
+    }
+}
+
+struct LayoutBuilder {
+    /// Always at least one element.
+    fields: Vec<MemcpyField>,
+}
+
+impl Default for LayoutBuilder {
+    fn default() -> Self {
+        Self {
+            fields: vec![MemcpyField {
+                bflatn_offset: 0,
+                bsatn_offset: 0,
+                length: 0,
+            }],
+        }
+    }
+}
+
+impl LayoutBuilder {
+    fn build(self) -> KnownBsatnLayout {
+        let LayoutBuilder { fields } = self;
+        let fields: Vec<_> = fields.into_iter().filter(|field| !field.is_empty()).collect();
+        let bsatn_length = fields.last().map(|last| last.bsatn_offset + last.length).unwrap_or(0);
+        KnownBsatnLayout { bsatn_length, fields }
+    }
+
+    fn current_field(&self) -> &MemcpyField {
+        self.fields.last().unwrap()
+    }
+
+    fn current_field_mut(&mut self) -> &mut MemcpyField {
+        self.fields.last_mut().unwrap()
+    }
+
+    fn next_bflatn_offset(&self) -> u16 {
+        let last = self.current_field();
+        last.bflatn_offset + last.length
```

Comment:
No, padding can only be introduced by products, so we only have to handle it in `visit_product`/`visit_product_element`. I expanded the comment in `visit_product_element`:
```rust
// Note that this is the only place we have to reason about alignment and padding
// because the enclosing `ProductTypeLayout` has already computed valid aligned offsets
// for the elements.
```